File: bindings/java/src/main/java/com/xfa/pdf/RenderedImage.java
package com.xfa.pdf;

import java.awt.image.BufferedImage;
import java.nio.ByteBuffer;

/**
 * A rendered page image with RGBA pixel data.
 */
public class RenderedImage {
    private final int width;
    private final int height;
    private final byte[] pixels;

    RenderedImage(int width, int height, byte[] pixels) {
        this.width = width;
        this.height = height;
        this.pixels = pixels;
    }

    /** Width in pixels. */
    public int getWidth() {
        return width;
    }

    /** Height in pixels. */
    public int getHeight() {
        return height;
    }

    /** Raw RGBA pixel data (4 bytes per pixel, row-major). */
    public byte[] getPixels() {
        return pixels;
    }

    /**
     * Convert to a BufferedImage (TYPE_INT_ARGB).
     * Requires java.awt — not available on Android.
     */
    public BufferedImage toBufferedImage() {
        BufferedImage img = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int offset = (y * width + x) * 4;
                int r = pixels[offset] & 0xFF;
                int g = pixels[offset + 1] & 0xFF;
                int b = pixels[offset + 2] & 0xFF;
                int a = pixels[offset + 3] & 0xFF;
                int argb = (a << 24) | (r << 16) | (g << 8) | b;
                img.setRGB(x, y, argb);
            }
        }
        return img;
    }
}

File: bindings/android/xfapdf/src/main/java/com/xfa/pdf/RenderedImage.kt
package com.xfa.pdf

import android.graphics.Bitmap
import java.nio.ByteBuffer

/**
 * An RGBA-rendered page image.
 *
 * @property width Image width in pixels.
 * @property height Image height in pixels.
 * @property pixels Raw RGBA pixel data (4 bytes per pixel).
 */
class RenderedImage internal constructor(
    val width: Int,
    val height: Int,
    val pixels: ByteArray
) {
    /**
     * Convert to an Android [Bitmap].
     *
     * The returned Bitmap uses [Bitmap.Config.ARGB_8888] format.
     */
    fun toBitmap(): Bitmap {
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        // RGBA -> ARGB conversion: Android expects ARGB_8888
        val argbPixels = IntArray(width * height)
        for (i in argbPixels.indices) {
            val offset = i * 4
            val r = pixels[offset].toInt() and 0xFF
            val g = pixels[offset + 1].toInt() and 0xFF
            val b = pixels[offset + 2].toInt() and 0xFF
            val a = pixels[offset + 3].toInt() and 0xFF
            argbPixels[i] = (a shl 24) or (r shl 16) or (g shl 8) or b
        }
        bitmap.setPixels(argbPixels, 0, width, 0, 0, width, height)
        return bitmap
    }
}
